cli/tools/jupyter/server.rs

```diff
@@ -0,0 +1,351 @@
+        }
+        _ => {
+          eprintln!(
+            "Unrecognized control message type: {}",
+            msg.message_type()
+          );
+        }
+      }
+    }
+  }
+
+  async fn handle_shell(
+    &mut self,
+    mut connection: Connection<zeromq::RouterSocket>,
+  ) -> Result<(), AnyError> {
+    loop {
+      let msg = JupyterMessage::read(&mut connection).await?;
+      self.handle_shell_message(msg, &mut connection).await?;
+    }
+  }
+
+  async fn handle_shell_message(
+    &mut self,
+    msg: JupyterMessage,
+    connection: &mut Connection<zeromq::RouterSocket>,
+  ) -> Result<(), AnyError> {
+    msg
+      .new_message("status")
+      .with_content(json!({"execution_state": "busy"}))
+      .send(&mut *self.iopub_socket.lock().await)
+      .await?;
+
+    match msg.message_type() {
+      "kernel_info_request" => {
+        msg
+          .new_reply()
+          .with_content(kernel_info())
+          .send(connection)
+          .await?;
+      }
+      "is_complete_request" => {
+        msg
+          .new_reply()
+          .with_content(json!({"status": "complete"}))
+          .send(connection)
+          .await?;
+      }
+      "execute_request" => {
+        self
+          .handle_execution_request(msg.clone(), connection)
+          .await?;
+      }
+      "comm_open" => {
+        msg
+          .comm_close_message()
+          .send(&mut *self.iopub_socket.lock().await)
+          .await?;
+      }
+      "complete_request" => todo!(),
+      "comm_msg" | "comm_info_request" | "history_request" => {
+        // We don't handle these messages
+      }
+      _ => {
+        eprintln!("Unrecognized shell message type: {}", msg.message_type());
+      }
+    }
+
+    msg
+      .new_message("status")
+      .with_content(json!({"execution_state": "idle"}))
+      .send(&mut *self.iopub_socket.lock().await)
+      .await?;
+    Ok(())
+  }
+
+  async fn handle_execution_request(
+    &mut self,
+    msg: JupyterMessage,
+    connection: &mut Connection<zeromq::RouterSocket>,
+  ) -> Result<(), AnyError> {
+    self.execution_count += 1;
+    *self.last_execution_request.borrow_mut() = Some(msg.clone());
+
+    msg
+      .new_message("execute_input")
+      .with_content(json!({
+          "execution_count": self.execution_count,
+          "code": msg.code()
+      }))
+      .send(&mut *self.iopub_socket.lock().await)
+      .await?;
+
+    let evaluate_response = self
+      .repl_session
+      .evaluate_line_with_object_wrapping(msg.code())
+      .await?;
+
+    let repl::cdp::EvaluateResponse {
+      result,
+      exception_details,
+    } = evaluate_response.value;
+
+    if exception_details.is_none() {
+      let output = self.repl_session.get_eval_value(&result).await?;
+      msg
+        .new_message("execute_result")
+        .with_content(json!({
+            "execution_count": self.execution_count,
+            "data": {
+                "text/plain": output
```

Comment:
Once this first draft is finished, I'd love to talk object representations and some kind of API to be able to send `display_data` messages (rich media types like HTML, json, etc.).

One thing I'd love to play with, once I have this branch working is try out returning `application/json` of `result.value` if it's an object (rather than a primitive).

Reply:
The original iteration of this PR had this capability - there was `Deno.jupyter` JavaScript API that had methods like `displayHtml` or `displayVega`. I don't feel comfortable adding such APIs for know though. I think we might be better off by leveraging API like `Deno.inspect()` that could provide nicer formatting. But I'm curious to hear opinions :)

Reply:
Yeah I'm not sure I'd want to support a wide range of options at this point. What might be the best focus is to provide enough tooling for library authors to opt in to using some more raw form like this:

```
obj  = {
  'text/html': "<b>test</b>"
}

opts = {
  raw: true
}

Deno.jupyter.display(obj, opts)
```

In IPython land as I'm sure you already know, there's a global `display` function (coming from `IPython.core.display_functions`) that provides enough building blocks to display any object

```python
import pandas as pd
df = pd.DataFrame({
    'Dinosaur': ['T-Rex', 'Velociraptor', 'Stegosaurus'],
    'Period': ['Cretaceous', 'Cretaceous', 'Jurassic'],
    'Length_meters': [12, 2, 9]
})
display(df)
```

or do a raw display call specifying what media types are available:

```
display({'text/html': "<b>test</b>"}, raw=True)
```

Based on what I've read so far, `Deno.inspect` _basically_ gives us a `__repr__` equivalent since it has a `Symbol` you can place on an object. That's both powerful and a bit limiting since Deno would need to be a dependency for every library where we'd want them to provide rich representations (tables, visualization, etc.).

In Python any library author can add methods like `__repr__` for the `text/plain` representation, `_repr_markdown_` for simple markdown, `_repr_html_`, and many other media types. When objects get inspected by IPython to create an `execute_result` or a `display_data` output, IPython checks for these methods. 

The spark that rose IPython and Jupyter notebooks to popularity was that library authors (e.g. Pandas and Matplotlib) embraced adding just one method to their classes and we immediately had visualizations of tables and plots inline. The rest is history from there.

The way I'd prioritize this would be to create a super low level way of displaying a bundle of media types (referred to as a mimebundle in jupyter docs) and a way to override (?) the way that an object is "formatted" when inspected (some function that goes from an arbitrary javascript object to that media bundle).

Reply:
Discussed on Discord, I have a plan how we can address this for userland libraries to be able to leverage and not depend on `Deno` itself. Will do a prototype before landing.